src/sql/migrations.rs

```diff
@@ -501,6 +501,40 @@ item TEXT DEFAULT '');"#,
         sql.execute_migration("ALTER TABLE msgs ADD COLUMN hop_info TEXT;", 81)
             .await?;
     }
+    if dbversion < 82 {
+        info!(context, "[migration] v82");
+        sql.execute_migration(
+            r#"CREATE TABLE imap (
+id INTEGER PRIMARY KEY AUTOINCREMENT,
+rfc724_mid TEXT DFEAULT '', -- Message-ID header
+folder TEXT DEFAULT '', -- IMAP folder
+target TEXT DEFAULT '', -- Destination folder, empty to delete.
+uid INTEGER DEFAULT 0, -- UID
+uidvalidity INTEGER DEFAULT 0,
+UNIQUE (folder, uid, uidvalidity)
+);
+CREATE INDEX imap_folder ON imap(folder);
+CREATE INDEX imap_messageid ON imap(rfc724_mid);
+
+INSERT INTO imap
+(rfc724_mid, folder, target, uid, uidvalidity)
+SELECT
+rfc724_mid,
+server_folder AS folder,
+server_folder AS target,
+server_uid AS uid,
+(SELECT uidvalidity FROM imap_sync WHERE folder=server_folder) AS uidvalidity
+FROM msgs
+WHERE server_uid>0
+ON CONFLICT (folder, uid, uidvalidity)
+DO UPDATE SET rfc724_mid=excluded.rfc724_mid,
+              target=excluded.target;
```

Comment:
It is the upsert clause: https://www.sqlite.org/lang_UPSERT.html
Without this, if `msgs` table contained two messages with the same `server_folder` and `server_uid` pair, `INSERT` will fail due to uniqueness constraint. Unfortunately `msgs` does not contain corresponding UID validity, so this can easily happen if UID validity was changed at some point, because existing DC versions do not clear `uid` column for `msgs` whenever UID validity changes. This clause ensures that `INSERT` never fails due to uniqueness constraint, but instead updates existing row by setting `rfc724_mid` and `target` to the values from "excluded" row (the one that was not inserted).

Reply:
oh, thanks for the explanation and docs pointer.  TIL